articles/files-response/src/app.controller.ts

```diff
@@ -1,12 +1,40 @@
-import { Controller, Get } from '@nestjs/common';
-import { AppService } from './app.service';
+import { Controller, Get, Header, Res } from '@nestjs/common';
+import { Response } from 'express';
+import { LibraryService } from './library/library.service';
+import { BookToCsvTransformer } from './dto/book-to-csv.transformer';
 
 @Controller()
 export class AppController {
-  constructor(private readonly appService: AppService) {}
+  constructor(private readonly libraryService: LibraryService) {}
 
+  @Header('Content-Type', 'text/csv')
   @Get()
-  getHello(): string {
-    return this.appService.getHello();
+  async getBooks(@Res() response: Response) {
+    const data = (await this.libraryService.getAll()).pipe(
+      new BookToCsvTransformer(),
+    );
+    data.on('data', (chunk) => {
+      response.write(chunk);
+    });
+    data.on('close', () => {
+      response.end();
+    });
+    data.on('error', () => {
+      response.status(500).send();
+    });
+  }
+
+  @Get('books.csv')
+  async getBooksFromFile(@Res() response: Response) {
+    const data = await this.libraryService.getFromStorage();
+    data.on('data', (chunk) => {
```

Comment:
why not just data.pipe(response) ?

Reply:
as it is an example, we can skip error handling to simplify it. Should we? @Sikora00

Reply:
You can leave the error part, but pipe instead of on 'data' has the ability to prevent backpressure.